src/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1175,7 +1175,27 @@ size_t QgsOgrProvider::layerCount() const
  */
 QgsWkbTypes::Type QgsOgrProvider::wkbType() const
 {
-  return static_cast<QgsWkbTypes::Type>( mOGRGeomType );
+  OGRwkbGeometryType wkb = mOGRGeomType;
+
+  if ( ogrDriverName == QLatin1String( "ESRI Shapefile" ) )
+  {
+    switch ( mOGRGeomType )
```

Comment:
I'd leave the shapefile check (unless that's something to fix upstream in gdal, @rouault?).